sn_interface/src/network_knowledge/sections_dag.rs

```diff
@@ -50,17 +53,87 @@ impl Sha3Hash for SectionInfo {
 }
 
 /// A Merkle DAG of BLS keys where every key is signed by its parent key, except the genesis one.
-#[derive(Clone, PartialEq, Eq, Deserialize, Serialize)]
+#[derive(Clone, PartialEq, Eq)]
 pub struct SectionsDAG {
     genesis_key: bls::PublicKey,
     dag: MerkleReg<SectionInfo>,
     dag_root: BTreeSet<bls::PublicKey>,
     hashes: BTreeMap<bls::PublicKey, Hash>,
 }
 
+impl Serialize for SectionsDAG {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut sections: Vec<(bls::PublicKey, SectionInfo)> = Vec::new();
+        let mut need_to_visit: Vec<(bls::PublicKey, Node<SectionInfo>)> = Vec::new();
+        for key in self.dag_root.iter() {
+            need_to_visit.push((
+                *self.genesis_key(),
+                self.get_node(key).map_err(S::Error::custom)?,
+            ));
+        }
+        while let Some((parent_key, current_node)) = need_to_visit.pop() {
+            need_to_visit.extend(
+                self.child_nodes(current_node.hash())
+                    .into_iter()
+                    .map(|child_node| (current_node.value.key, child_node)),
+            );
+            sections.push((parent_key, current_node.value));
+        }
+
+        let mut map = serializer.serialize_map(Some(2))?;
+        map.serialize_entry("genesis_key", &self.genesis_key)?;
+        map.serialize_entry("sections", &sections)?;
+        map.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for SectionsDAG {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct Intermediate {
+            genesis_key: bls::PublicKey,
+            sections: Vec<(bls::PublicKey, SectionInfo)>,
+        }
+        let inter: Intermediate = Deserialize::deserialize(deserializer)?;
+
+        let mut dag = SectionsDAG::new(inter.genesis_key);
+        for (parent, info) in inter.sections {
+            dag.insert(&parent, info.key, info.sig)
+                .map_err(D::Error::custom)?;
+        }
+        Ok(dag)
+    }
+}
+
 impl Debug for SectionsDAG {
     fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
-        write!(formatter, "{:?}", self.keys().format(", "))
+        let keys: Vec<_> = self.keys().cloned().collect();
+        let mut parent_indices: Vec<usize> = Vec::new();
+        for key in keys.iter() {
+            match self.get_parent_key(key).map_err(|_| fmt::Error)? {
+                None => parent_indices.push(usize::MAX),
+                Some(parent_key) => {
+                    let idx = keys
+                        .iter()
+                        .position(|key| *key == parent_key)
+                        .ok_or(fmt::Error)?;
+                    parent_indices.push(idx)
+                }
+            }
+        }
```

Comment:
That parent key position() call will make this debug impl quadratic as well, and I imagine we may be logging this quite frequently. Lets pre-index the keys to avoid any slowdown.
```suggestion
        let keys: Vec<_> = self.keys().cloned().collect();
        let key_positions: BTreeMap<&PublicKey, usize> = keys.iter().enumerate().collect();
        let mut parent_indices: Vec<usize> = Vec::new();
        for key in keys.iter() {
            match self.get_parent_key(key).map_err(|_| fmt::Error)? {
                None => parent_indices.push(usize::MAX),
                Some(parent_key) => {
                    let idx = key_positions.get(&parent_key).ok_or(fmt::Error)?;
                    parent_indices.push(idx)
                }
            }
        }
```

Reply:
Updated it!